dodiscover/_protocol.py

```diff
@@ -0,0 +1,52 @@
+from typing import Dict, FrozenSet, Iterable, Protocol
+
+import networkx as nx
+
+
+class GraphProtocol(Protocol):
```

Comment:
Interesting.  What is the rationale for using duck typing instead of subclassing?

Reply:
Protocols just are Python's way of defining an abstract interface. So in the code, note that `pywhy-graphs` is never imported. And does not need to be. This is in line w/ discussions we had awhile back where the Dev team requested that we don't explicitly have to rely on pywhy's implementation of causal graphs.

This `GraphProtocol` class then allows a user to define their own graph class, as long as it adheres to the API, then it "should" still work. In addition, this provides `mypy` type checking w/o having to import `pywhy-graphs` explicitly.

Reply:
Just to add an explicit example:

if someone from say `ananke` wants to make their graph implementation work with the implementation of the algorithms here that rely on a GraphProtocol, then they just need to modify their API to have the functions specified.

I personally am okay w/ being opinionated and requiring users to use pywhy-graphs directly and providing a transformation module for converting graphs from common formats, but I think we can always swap that easily using ctrl+f+replace